File: Leetcode/easy/ReverseLinkedList.scala
// Given the head of a singly linked list, reverse the list, and return the reversed list.
// Time O(n), Space O(1)

object Solution {
    def reverseList(head: ListNode): ListNode = {
        @annotation.tailrec
        def reverseRec(prev: ListNode, cur: ListNode, next: ListNode): ListNode = {
            cur.next = prev
            if (next != null) reverseRec(cur, next, next.next)
            else cur
        }

        if (head == null) head
        else reverseRec(null, head, head.next)
    }
}

File: Leetcode/easy/ReverseLinkedList.js
// Given the head of a singly linked list, reverse the list, and return the reversed list.
// Time O(n), Space O(1)

const reverseList = function (head) {
    let prev = null;

    while (head) {
        let next = head.next;
        head.next = prev;
        prev = head;
        head = next;
    }
    return prev;
};